common/remote/hooks/azure/azurefile.cpp

```diff
@@ -0,0 +1,827 @@
+    virtual unsigned getCRC() { UNIMPLEMENTED; }
+    virtual void setCreateFlags(unsigned short cflags) { UNIMPLEMENTED; }
+    virtual void setShareMode(IFSHmode shmode) { UNIMPLEMENTED; }
+    virtual bool createDirectory();
+    virtual IDirectoryDifferenceIterator *monitorDirectory(
+                                  IDirectoryIterator *prev=NULL,    // in (NULL means use current as baseline)
+                                  const char *mask=NULL,
+                                  bool sub=false,
+                                  bool includedirs=false,
+                                  unsigned checkinterval=60*1000,
+                                  unsigned timeout=(unsigned)-1,
+                                  Semaphore *abortsem=NULL)  { UNIMPLEMENTED; }
+    virtual void copySection(const RemoteFilename &dest, offset_t toOfs=(offset_t)-1, offset_t fromOfs=0, offset_t size=(offset_t)-1, ICopyFileProgress *progress=NULL, CFflags copyFlags=CFnone) { UNIMPLEMENTED; }
+    virtual void copyTo(IFile *dest, size32_t buffersize=DEFAULT_COPY_BLKSIZE, ICopyFileProgress *progress=NULL, bool usetmp=false, CFflags copyFlags=CFnone) { UNIMPLEMENTED; }
+    virtual IMemoryMappedFile *openMemoryMapped(offset_t ofs=0, memsize_t len=(memsize_t)-1, bool write=false)  { UNIMPLEMENTED; }
+
+protected:
+    void createAppendBlob();
+    void appendToAppendBlob(size32_t len, const void * data);
+    void createBlockBlob();
+    void appendToBlockBlob(size32_t len, const void * data);
+    void commitBlockBlob();
+
+    offset_t readBlock(BlobPageContents & contents, FileIOStats & stats, offset_t from = 0, offset_t length = unknownFileSize);
+    void ensureMetaData();
+    void gatherMetaData();
+    IFileIO * createFileReadIO();
+    IFileIO * createFileWriteIO();
+    std::shared_ptr<azure::storage_lite::blob_client> getClient();
+
+protected:
+    StringBuffer fullName;
+    StringAttr accountName;
+    StringAttr accountKey;
+    std::string containerName;
+    std::string blobName;
+    offset_t fileSize = unknownFileSize;
+    bool haveMeta = false;
+    bool isDir = false;
+    bool fileExists = false;
+    time_t blobModifiedTime = 0;
+    SpinLock lock;
+};
+
+
+//---------------------------------------------------------------------------------------------------------------------
+
+AzureFileReadIO::AzureFileReadIO(AzureFile * _file, const FileIOStats & _firstStats)
+: file(_file), stats(_firstStats)
+{
+    startResultOffset = 0;
+    endResultOffset = 0;
+}
+
+size32_t AzureFileReadIO::read(offset_t pos, size32_t len, void * data)
+{
+    if (pos > file->fileSize)
+        return 0;
+    if (pos + len > file->fileSize)
+        len = file->fileSize - pos;
+    if (len == 0)
+        return 0;
+
+    size32_t sizeRead = 0;
+    offset_t lastOffset = pos + len;
+
+    // MORE: Do we ever read file IO from more than one thread?  I'm not convinced we do, and the critical blocks waste space and slow it down.
```

Comment:
I would think (at least in Roxie) that fetch and index reads may well end up reading a shared ifileio from multiple threads.

Reply:
The code currently protects for multi threaded access.  If we swap to a streaming interface then the cs shouldn't be necessary.  Can revisit when that works is done.